File: proposed-projects/round-trip-experiment/reverse/20260909T004354Z_a47a3057/20260910T044103Z_5267f914/ported_implementation/gbnf/grammar_graph/print.py
from typing import TYPE_CHECKING, Callable, List

from .colorize import Color
from .get_parent_stack_id import get_parent_stack_id
from .grammar_graph_types import PrintOpts
from .type_guards import is_range, is_rule_char, is_rule_ref

if TYPE_CHECKING:  # pragma: no cover - typing only
    from .graph_node import GraphNode
    from .graph_pointer import GraphPointer


def get_char(char_code: int) -> str:
    char = chr(char_code)
    if char == "\n":
        return "\\n"
    return char


def print_graph_pointer(pointer: "GraphPointer") -> Callable[[PrintOpts], str]:
    def print_pointer(opts: PrintOpts) -> str:
        col = opts.colorize
        return col(f"*{get_parent_stack_id(pointer, col)}", Color.RED)

    return print_pointer


def print_graph_node(node: "GraphNode") -> Callable[[PrintOpts], str]:
    def print_node(opts: PrintOpts) -> str:
        pointers = opts.pointers
        col = opts.colorize
        show_position = opts.show_position
        rule = node.rule
        parts: List[str] = []
        if show_position:
            parts.extend(
                [
                    col("{", Color.BLUE),
                    col(node.id, Color.GRAY),
                    col("}", Color.BLUE),
                ]
            )

        if is_rule_char(rule):
            parts.extend(
                [
                    col("[", Color.GRAY),
                    col(
                        "".join(
                            "".join(col(get_char(val), Color.YELLOW) for val in v)
                            if is_range(v)
                            else get_char(v)
                            for v in rule.value
                        ),
                        Color.YELLOW,
                    ),
                    col("]", Color.GRAY),
                ]
            )
        elif is_rule_ref(rule):
            parts.extend(
                [
                    col("Ref(", Color.GRAY),
                    col(rule.value, Color.GREEN),
                    col(")", Color.GRAY),
                ]
            )
        else:
            parts.append(col(rule.type, Color.YELLOW))

        if pointers is not None and pointers.size > 0:
            for pointer in pointers:
                pointer_parts: List[str] = []
                if pointer.node is node:
                    pointer_parts.append(pointer.print(opts))

                if len(pointer_parts) > 0:
                    parts.extend(
                        [
                            col("[", Color.GRAY),
                            col("".join(pointer_parts), Color.YELLOW),
                            col("]", Color.GRAY),
                        ]
                    )

        parts_to_return = ["".join(parts)]
        if node.next:
            parts_to_return.append(node.next.print(opts))

        return col("-> ", Color.GRAY).join(parts_to_return)

    return print_node

File: packages/gbnf-experiment/data/20260909T002206Z_273cf6b9/ported_implementation/src/grammar-graph/print.ts
import { Color } from './colorize.ts';
import { getParentStackId } from './get-parent-stack-id.ts';
import type { PrintOpts } from './grammar-graph-types.ts';
import type { GraphNode } from './graph-node.ts';
import type { GraphPointer } from './graph-pointer.ts';
import { isRange, isRuleChar, isRuleRef } from './type-guards.ts';

export const printGraphPointer = (pointer: GraphPointer) => (opts: PrintOpts): string => {
  const col = opts.colorize;
  return col(
    `*${getParentStackId(pointer, col)}`,
    Color.RED,
  );
};

export const printGraphNode = (node: GraphNode) => (opts: PrintOpts): string => {
  const pointers = opts.pointers;
  const col = opts.colorize;
  const showPosition = opts.show_position ?? false;
  const rule = node.rule;
  const parts: string[] = [];
  if (showPosition) {
    parts.push(col('{', Color.BLUE), col(node.id, Color.GRAY), col('}', Color.BLUE));
  }

  if (isRuleChar(rule)) {
    parts.push(
      col('[', Color.GRAY),
      col(
        rule.value.map(v => (
          isRange(v)
            ? v.map(val => col(getChar(val), Color.YELLOW)).join('')
            : getChar(v)
        )).join(''),
        Color.YELLOW,
      ),
      col(']', Color.GRAY),
    );
  } else if (isRuleRef(rule)) {
    parts.push(
      col('Ref(', Color.GRAY),
      col(rule.value, Color.GREEN),
      col(')', Color.GRAY),
    );
  } else {
    parts.push(col(rule.type, Color.YELLOW));
  }

  if (pointers && pointers.size) {
    for (const pointer of pointers) {
      const pointerParts: string[] = [];
      if (pointer.node === node) {
        pointerParts.push(pointer.print(opts));
      }

      if (pointerParts.length > 0) {
        parts.push(
          col('[', Color.GRAY),
          col(pointerParts.join(''), Color.YELLOW),
          col(']', Color.GRAY),
        );
      }
    }
  }

  const partsToReturn = [
    parts.join(''),
  ];
  if (node.next) {
    partsToReturn.push(node.next.print(opts));
  }

  return partsToReturn.join(col('-> ', Color.GRAY));
};

export const getChar = (charCode: number): string => {
  const char = String.fromCodePoint(charCode);
  if (char === '\n') {
    return '\\n';
  }
  return char;
};
